teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
 a managed Tekton, users can be prevented access to the `tekton-pipelines`
 namespaces, where the feature flag config map is hosted.
 
+## Proposal
+
+We will tackle separately features that are controlled via new fields in the API
+and features that are only behavioral. (The alternative
+[one feature flag per new alpha field](#one-feature-flag-per-new-alpha-field) describes
+an alternative where we use the same approach for both.)
+
+In both cases we want to control access with
+[a feature flag](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior).
+
+* For new API fields, will will gate access to all of them behind two feature flags
+  called `enable-alpha-api-fields` and `enable-beta-api-fields`.
+  ([Design details: new api fields](#new-api-fields).)
+* For changes in behavior which are not configurable via API fields in our CRDs,
+  each feature will be gated by its own unique feature flag.
+  ([Design details: changes in behavior](#changes-in-behavior).)
+
+
+### Risks and Mitigations
+
+See the pros and cons sections in the [design details](#design-details)
+below.
+
+## Design Details
+
+### New API fields
+
+We gate access to all new API fields with two new
+[feature flags](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior) called `enable-alpha-api-fields` and `enable-beta-api-fields`
+which will default to false.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  enable-alpha-api-fields: "true"
+  enable-beta-api-fields: "true"
```

Comment:
After reading and re-reading this a few times, I'm starting to lean towards using a different flag (the name is not good, sorry):
```
data:
 extra-api-fields: ["current" | "alpha" | "beta" ]
```

Where "current" means only the API fields that are on by default in the current API, "beta" means enable "beta" fields only in the current API, "alpha" means enabled "alpha" and "beta" fields in the current API.

What I'm trying to avoid is the case of someone enabling "alpha" fields and disabling "beta" ones. 
If a user / operator wants to stay to up date with the latest and greatest features from Tekton, then they will enable "alpha" level features. If they want to have a single feature that is in alpha and do not want anything else, then this solution won't work for them anyways, because it does not allow to pick specific features.

Reply:
I tend to like that approach (except the name `current` 😛 ).

> Once we have v1, will v1alpha and v1beta be frozen and eventually removed? Will new fields only ever be added to v1 as alpha fields first?

Yes, existing types (`Pipeline`, …) would have `v1alpha1` and `v1beta1` "frozen" and removed at some point. Any new field on those existing type would happen on the `v1`. Any **now type** though, could go through the `v1alpha1` -> `v1beta1` -> `v1` flow.

Reply:
That's way better @afrittoli ! Thank you!

How about this naming:

`enable-api-fields: ["stable" | "alpha" | "beta" ]`

"Stable" still isn't perfect but I think it's okay? Another option would be to initially add only:

`enable-api-fields: [ "alpha" | "beta" ]` with `beta` being the default (i.e. the greatest stability level we support), and when we move to v1 we'd update it to be `enable-api-fields: [ "v1" | "alpha" | "beta" ]` with v1 as the default?

Reply:
this simplifies the tests we'll need to run as well! just one set for each value of the flag (i.e. vs. all the combinations... okay maybe that just saves us one but STILL!) 🎉